openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/schematypes/ReferencedSchemaType.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.openapi.generators.schema.schematypes;
+
+import io.ballerina.compiler.syntax.tree.TypeDescriptorNode;
+import io.ballerina.openapi.exception.BallerinaOpenApiException;
+import io.ballerina.openapi.generators.schema.SchemaUtils;
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.media.Schema;
+
+import static io.ballerina.compiler.syntax.tree.NodeFactory.createIdentifierToken;
+import static io.ballerina.compiler.syntax.tree.NodeFactory.createSimpleNameReferenceNode;
+import static io.ballerina.openapi.generators.GeneratorUtils.extractReferenceType;
+import static io.ballerina.openapi.generators.GeneratorUtils.getValidName;
+
+/**
+ * Generate TypeDefinitionNode and TypeDescriptorNode for referenced schemas.
+ *
+ * @since 2.0.0
+ */
+public class ReferencedSchemaType extends SchemaType {
+
+    private final OpenAPI openAPI;
+    private final boolean nullable;
+
+    public ReferencedSchemaType(OpenAPI openAPI, boolean nullable) {
+        this.openAPI = openAPI;
+        this.nullable = nullable;
+    }
+
+    /**
+     * Generate TypeDescriptorNode for referenced schemas.
+     */
+    @Override
+    public TypeDescriptorNode generateTypeDescriptorNode(Schema schema) throws BallerinaOpenApiException {
+        String typeName = getValidName(extractReferenceType(schema.get$ref()), true);
+        Schema<?> refSchema = openAPI.getComponents().getSchemas().get(typeName);
```

Comment:
@indikasampath2000 Regarding the suggestion to not pass OpenAPI definition to `SchemaType` sub classes. 
```yaml
Pet:
      required:
        - id
        - name
      nullable: true
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        tag:
          type: string
        type:
          type: string
    Dog: 
      $ref: "#/components/schemas/Pet"
```
`ReferencedSchemaType.generateTypeDescriptorNode` is getting called when creating a Ballerina type for the `Dog` schema. `Dog` is type of `Pet` and `Pet` schema is `nullable:true`. In the current implementation `Dog` ballerina type will be created as below where Dog is type `Pet` or `nill`. 

`public type Dog Pet?;`

To check whether the reference schema is `nullable:true` open API definition should be available at this class. 

We also have an option to make a record it self nillable by adding `?` after the record definition as below. 

```ballerina
public type Pet record {
    int id;
    string name;
    string tag?;
    string 'type?;
}?;
```
This means `Pet` can be a record or nill. However if we use this approach we need to be extra careful at record inclusion.